src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp

```diff
@@ -16,112 +15,138 @@
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
 
-            //BansheeCurse
-            if (BansheeCurse_Timer <= diff)
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
             {
-                if (rand32() % 100 < 75)
+                case EVENT_SPELL_BANSHEEWAIL:
+                    DoCastVictim(SPELL_BANSHEEWAIL);
+                    _events.Repeat(4s);
+                    break;
+                case EVENT_SPELL_BANSHEECURSE:
                     DoCastVictim(SPELL_BANSHEECURSE);
-                //18 seconds until we should cast this again
-                BansheeCurse_Timer = 18000;
-            } else BansheeCurse_Timer -= diff;
-
-            //Silence
-            if (Silence_Timer <= diff)
-            {
-                if (rand32() % 100 < 80)
+                    _events.Repeat(18s);
+                    break;
+                case EVENT_SPELL_SILENCE:
                     DoCastVictim(SPELL_SILENCE);
-                //13 seconds until we should cast this again
-                Silence_Timer = 13000;
-            } else Silence_Timer -= diff;
-
-            DoMeleeAttackIfReady();
+                    _events.ScheduleEvent(EVENT_SPELL_SILENCE, 13s);
+                    break;
+                case EVENT_SPELL_POSSESS:
+                    if (Unit* possessTarget = SelectTarget(SELECT_TARGET_RANDOM, 1, 0, true, false))    // Random target to be possessed
+                    {
+                        if (possessTarget->IsAlive())
+                        {
+                            me->CastStop();
+                            DoCast(possessTarget, SPELL_POSSESS);
+                            DoCast(possessTarget, SPELL_POSSESSED);
+                            DoCastSelf(SPELL_POSSESS_INV);
+                            _possessedTargetGuid = possessTarget->GetGUID();
+                            _events.ScheduleEvent(EVENT_CHECK_POSSESSED, 0s);
+                        }
+                        else
+                            _events.ScheduleEvent(EVENT_SPELL_POSSESS, 20s, 30s);
+                    }
+                    break;
+                case EVENT_INVISIBLE:
+                    if (Player* possessedTarget = ObjectAccessor::GetPlayer(*me, _possessedTargetGuid))    // When there's a possessed target
+                    {
+                        possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESS);
+                        possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESSED);
+                        me->RemoveAurasDueToSpell(SPELL_POSSESS_INV);
+                        possessedTarget->SetFullHealth();
+                        _possessedTargetGuid.Clear();
+                        _events.ScheduleEvent(EVENT_SPELL_POSSESS, 20s, 30s);
+                    }
+                    break;
+                case EVENT_CHECK_POSSESSED:
+                    if (me->HasAura(SPELL_POSSESS_INV))
```

Comment:
why is this check necessary?

just call cancelevent when removing the aura, no need to check every tick whether you still have it...